### ISSUE TITLE
CA1508 reported incorrectly

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn-analyzers#5789]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/CA1508-reported-incorrectly/1618815)._

---
[severity:It's more difficult to complete my work]
I have the following piece of code:

```csharp
class Customer
{
        public string FirstName { get; set; }
        public string LastName { get; set; }
}
```

When passing Customer to a function and testing for both first name and last name being empty, I get CA1508 explaining that the length of the last name will always be zero, yet this is clearly not the case...

> warning CA1508: 'customer. LastName.Length == 0' is always 'true'. Remove or refactor the condition(s) to avoid dead code.

```csharp
static class GenerateWarning
{
     private static void ShowProblem(Customer customer)
     {
        if (customer. FirstName.Length == 0 && customer. LastName.Length == 0)
        {
        }
     }
}
```

See full sample available on [GitHub](https://github.com/epsitec/issue-vs2022-ca1508).

The issue only appears when `Customer` is defined in another project.

---
#### Feedback Bot on 12/22/2021, 07:21 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 1/4/2022, 00:07 AM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

---
#### Manish Vasani [MSFT] solved on 1/5/2022, 11:26 AM, 0 votes: 

<p>This issue has been fixed in CA analyzers shipping in .NET6 SDK! Install the most recent release from <a target='_blank' href="https://dotnet.microsoft.com/en-us/download/dotnet/6.0">https://dotnet.microsoft.com/en-us/download/dotnet/6.0</a>. Thank you for providing valuable feedback which has helped improve the product.</p>

#### Pierre Arnaud on 1/5/2022, 00:37 PM: 

<p>I can confirm that this worked with <code>6.0.100</code> – however, with the latest SDK <code>6.0.101</code> or with the latest Visual Studio 2022 17.1.0 Preview 1.1 the <strong>warning CA1508 is present</strong> as described.</p>
<hr>
<pre><code>❯ dotnet build

Welcome to .NET 6.0!
<span class="hljs-comment">---------------------</span>
SDK Version: 6.0.101

Telemetry
<span class="hljs-comment">---------</span>
The .NET tools collect usage data in order to <span class="hljs-keyword">help</span> us improve your experience. It <span class="hljs-keyword">is</span> collected <span class="hljs-keyword">by</span> Microsoft <span class="hljs-keyword">and</span> <span class="hljs-keyword">shared</span> <span class="hljs-keyword">with</span> the community. You can opt-<span class="hljs-keyword">out</span> <span class="hljs-keyword">of</span> telemetry <span class="hljs-keyword">by</span> setting the DOTNET_CLI_TELEMETRY_OPTOUT environment <span class="hljs-keyword">variable</span> <span class="hljs-keyword">to</span> <span class="hljs-string">'1'</span> <span class="hljs-keyword">or</span> <span class="hljs-string">'true'</span> <span class="hljs-keyword">using</span> your favorite shell.

<span class="hljs-keyword">Read</span> more about .NET CLI Tools telemetry: https://aka.ms/<span class="hljs-keyword">dotnet</span>-cli-telemetry

<span class="hljs-comment">----------------</span>
Installed an ASP.NET Core HTTPS development certificate.
<span class="hljs-keyword">To</span> trust the certificate run <span class="hljs-string">'dotnet dev-certs https --trust'</span> (Windows <span class="hljs-keyword">and</span> macOS <span class="hljs-keyword">only</span>).
Learn about HTTPS: https://aka.ms/<span class="hljs-keyword">dotnet</span>-https
<span class="hljs-comment">----------------</span>
Write your <span class="hljs-keyword">first</span> app: https://aka.ms/<span class="hljs-keyword">dotnet</span>-hello-world
Find <span class="hljs-keyword">out</span> what<span class="hljs-string">'s new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use '</span><span class="hljs-keyword">dotnet</span> <span class="hljs-comment">--help' to see available commands or visit: https://aka.ms/dotnet-cli</span>
<span class="hljs-comment">--------------------------------------------------------------------------------------</span>
Microsoft (R) <span class="hljs-keyword">Build</span> <span class="hljs-keyword">Engine</span> <span class="hljs-keyword">version</span> <span class="hljs-number">17.0</span><span class="hljs-number">.0</span>+c9eb9dd64 <span class="hljs-keyword">for</span> .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects <span class="hljs-keyword">to</span> restore...
  Restored S:\git\tmp\issue-vs2022-ca1508\ClassLibrary\ClassLibrary.csproj (<span class="hljs-keyword">in</span> <span class="hljs-number">218</span> ms).
  Restored S:\git\tmp\issue-vs2022-ca1508\Program\Program.csproj (<span class="hljs-keyword">in</span> <span class="hljs-number">220</span> ms).
  ClassLibrary -&gt; S:\git\tmp\issue-vs2022-ca1508\ClassLibrary\<span class="hljs-keyword">bin</span>\Debug\net6<span class="hljs-number">.0</span>\ClassLibrary.dll
CSC : <span class="hljs-keyword">warning</span> CA1014: Mark assemblies <span class="hljs-keyword">with</span> CLSCompliant [S:\git\tmp\issue-vs2022-ca1508\Program\Program.csproj]
S:\git\tmp\issue-vs2022-ca1508\Program\GenerateWarning.cs(<span class="hljs-number">16</span>,<span class="hljs-number">14</span>): <span class="hljs-keyword">warning</span> CA1508: <span class="hljs-string">'customer.LastName.Length == 0'</span> <span class="hljs-keyword">is</span> <span class="hljs-keyword">always</span> <span class="hljs-string">'true'</span>. Remove <span class="hljs-keyword">or</span> refactor the condition(s) <span class="hljs-keyword">to</span> avoid dead code. [S:\git\tmp\issue-vs2022-ca1508\Program\Program.csproj]
  Program -&gt; S:\git\tmp\issue-vs2022-ca1508\Program\<span class="hljs-keyword">bin</span>\Debug\net6<span class="hljs-number">.0</span>\Program.dll

<span class="hljs-keyword">Build</span> succeeded.

CSC : <span class="hljs-keyword">warning</span> CA1014: Mark assemblies <span class="hljs-keyword">with</span> CLSCompliant [S:\git\tmp\issue-vs2022-ca1508\Program\Program.csproj]
S:\git\tmp\issue-vs2022-ca1508\Program\GenerateWarning.cs(<span class="hljs-number">16</span>,<span class="hljs-number">14</span>): <span class="hljs-keyword">warning</span> CA1508: <span class="hljs-string">'customer.LastName.Length == 0'</span> <span class="hljs-keyword">is</span> <span class="hljs-keyword">always</span> <span class="hljs-string">'true'</span>. Remove <span class="hljs-keyword">or</span> refactor the condition(s) <span class="hljs-keyword">to</span> avoid dead code. [S:\git\tmp\issue-vs2022-ca1508\Program\Program.csproj]
    <span class="hljs-number">2</span> <span class="hljs-keyword">Warning</span>(s)
    <span class="hljs-number">0</span> <span class="hljs-keyword">Error</span>(s)

<span class="hljs-built_in">Time</span> Elapsed <span class="hljs-number">00</span>:<span class="hljs-number">00</span>:<span class="hljs-number">06.08</span>
</code></pre>
<hr>
<pre><code>Rebuild started...
<span class="hljs-meta">1&gt;</span>------ Rebuild All started: Project: ClassLibrary, Configuration: Debug Any CPU ------
Restored S:\git\tmp\issue-vs2022-ca1508\ClassLibrary\ClassLibrary.csproj (in 111 ms).
Restored S:\git\tmp\issue-vs2022-ca1508\Program\Program.csproj (in 111 ms).
<span class="hljs-meta">1&gt;</span>ClassLibrary -&gt; S:\git\tmp\issue-vs2022-ca1508\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll
<span class="hljs-meta">2&gt;</span>------ Rebuild All started: Project: Program, Configuration: Debug Any CPU ------
<span class="hljs-meta">2&gt;</span>CSC : warning CA1014: Mark assemblies with CLSCompliant
<span class="hljs-meta">2&gt;</span>S:\git\tmp\issue-vs2022-ca1508\Program\GenerateWarning.cs(16,14,16,43): warning CA1508: <span class="hljs-string">'customer.LastName.Length == 0'</span> is always <span class="hljs-string">'true'</span>. Remove or refactor the condition(s) to avoid dead code.
<span class="hljs-meta">2&gt;</span>Program -&gt; S:\git\tmp\issue-vs2022-ca1508\Program\bin\Debug\net6.0\Program.dll
<span class="hljs-meta">2&gt;</span>Done building project <span class="hljs-string">"Program.csproj"</span>.
========== Rebuild All: 2 succeeded, 0 failed, 0 skipped ==========
</code></pre>
<p>Please reopen the issue.</p>

#### Catia Guidi on 1/5/2022, 04:02 PM: 

<p>I cloned the git project<br>
Downloaded the 6.0.101 SDK<br>
Executed a <code>dotnet clean</code> <code>dotnet build</code></p>
<p>And have the same warnings as mentioned</p>

#### Pierre Arnaud on 1/11/2022, 01:18 AM: 

<p>I’d like to add that with <code>17.1.0-preview-21610-01+96a618ca5</code> (aka <code>6.0.200-preview.21617.4</code>) installed on January 10 2022, the issue is still not resolved. I’d like to insist that this issue should be <strong>reopened</strong>.</p>

**Comment by mavasani** [comment]
NOTE: The issue repros only when the `Customer` type is defined in a separate project.

**Comment by epsitec** [comment]
Thank you @mavasani for reopening this topic and moving it to the `roslyn-analyzers` issues.

**Comment by Youssef1313** [comment]
@mavasani I can reproduce with the latest .NET 7 pre-release, but cannot reproduce in a unit test.

[CA1508.zip](https://github.com/dotnet/roslyn-analyzers/files/8014116/CA1508.zip)

Following is the test I wrote, which passes, meaning there are no produced diagnostics:

```csharp
        [Fact, WorkItem(5789, "https://github.com/dotnet/roslyn-analyzers/issues/5789")]
        public async Task TestTwoProjects()
        {
            var source = @"
public static class GenerateWarning
{
    public static void ShowProblem(Customer customer)
    {
        if ((customer.FirstName.Length == 0) &&
            (customer.LastName.Length == 0)) // warning CA1508: 'customer.LastName.Length == 0' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
        {
        }
    }
}
";
            var additionalProjectSource = @"
public sealed class Customer
{
    public string FirstName { get; set; }
    public string LastName { get; set; }
}
";
            await new VerifyCS.Test
            {
                TestState =
                {
                    Sources = { source },
                    AdditionalProjects =
                    {
                        ["OtherProject"] =
                        {
                            Sources =
                            {
                                ("Customer.cs", additionalProjectSource),
                            }
                        },
                    },
                    AdditionalProjectReferences = { "OtherProject" },
                },
            }.RunAsync();
        }
```

Any idea what I could have done wrong?

**Comment by epsitec** [comment]
I had this warning fire only if the code is not in the same as the class itself.

**Comment by mpidash** [comment]
I could not reproduce this bug with the preview SDK `8.0.100-preview.6.23330.14`. I tried the following SDKs:

Bug is reproducible with:

1. `6.0.412`
2. `7.0.306`
3. `8.0.100-preview.1.23115.2`

Bug is **NOT** reproducible with:

1. `8.0.100-preview.2.23157.25`
2. `8.0.100-preview.6.23330.14`

So this seems to be fixed in `8.0.100-preview.2.23157.25`.
